src/utils.jl

```diff
@@ -44,20 +44,76 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
 """
     reset_minibatch_train!(nlp::AbstractKnetNLPModel)
 
-Select a new training minibatch for `nlp`.
-Typically used before a new evaluation of the loss function/gradient.
+Select a the first training minibatch for `nlp`.
 """
 reset_minibatch_train!(nlp::AbstractKnetNLPModel) =
+  nlp.current_training_minibatch = first(nlp.training_minibatch_iterator)
+
+"""
+  rand_minibatch_train!(nlp::AbstractKnetNLPModel)
+
+Select a training minibatch for `nlp` randomly.
+"""
+rand_minibatch_train!(nlp::AbstractKnetNLPModel) =
   nlp.current_training_minibatch = rand(nlp.training_minibatch_iterator)
 
+"""
+minibatch_next_train!(nlp::AbstractKnetNLPModel)
+Selects the next mini-batch from training_minibatch_iterator
+nlp:: KnetNLPModel 
+i:: current location in the iterator
+Returns a new current location of the iterator i, if 0 is returned when all mini-batches are visited
+can be used in a loop or method call - refere to KnetNLPModelProblems for more use cases
+"""
+function minibatch_next_train!(nlp::AbstractKnetNLPModel, i::int)
+  i += nlp.size_minibatch # update the i by mini_batch size
```

Comment:
`nlp.size_minibatch` is an Integer telling in how many pieces the training dataset is divided.
I am not sure about the effect of this line.
Maybe you want something similar to 
```julia
i += length(nlp.training_data_set / nlp.size_minibatch 
```

Maybe you should add properly `i_training` and `i_test` in the `KnetNLPModel` structure.
This way you could have 
```julia
function minibatch_next_train!(nlp::AbstractKnetNLPModel; i::int=nlp.i)
```

Reply:
I agree with  i_training and i_test should be there  in the KnetNLPModel structure I will do the modification 

as a matter of fact I checked the  i += length(nlp.training_data_set / nlp.size_minibatch ) doesn't work 
The way knet does it is to ignore the size and do {i \in [1,data_size-batch_size]} which is terrible but that mean i for batch size of 1000 in MNIST is between 1 to 59999 which is weird for me but that is how it was written (I spent 2 days debugging till I find this weird behavior )

Reply:
check here the knet random code: https://github.com/denizyuret/Knet.jl/blob/e9bec84b826ee7231db4ca8772fae6eb2fd4bcb9/src/train20/data.jl

function rand(d::Data)
    i = rand(0:(d.length-d.batchsize))
    return iterate(d, i)[1]
end

Reply:
Did you try `length(nlp.training_data_set) / nlp.size_minibatch`, I forgot the coma in my previous answer.

```quote
that mean i for batch size of 1000 in MNIST is between 1 to 59999
```
I think the important point is: i <= training_minibatch_iterator.imax